app/models/user.rb

```diff
@@ -49,4 +74,44 @@ def self.authenticate_with_username_or_email(username_or_email,login_password)
   def to_s
     username
   end
+
+  # Activates an account.
+  def activate
+    update_attribute(:activated,    true)
+    update_attribute(:activated_at, Time.zone.now)
+  end
+
+  # Sends activation email.
+  def send_activation_email
+    UserMailer.account_activation(self).deliver_now
+  end
+  
+  # Sets the password reset attributes.
+  def create_reset_digest
+    self.reset_token = User.new_token
+    update_attribute(:reset_digest,  User.digest(reset_token))
+    update_attribute(:reset_sent_at, Time.zone.now)
+  end
+
+   # Sends password reset email.
+  def send_password_reset_email
+    UserMailer.password_reset(self).deliver_now
+  end
+
+  # Returns true if a password reset has expired.
+  def password_reset_expired?
+    reset_sent_at < 2.hours.ago
+  end
+
+private
+  # Converts email to all lower-case.
+  def downcase_email
+    self.email = email.downcase
```

Comment:
Why do you need `self`?

Also we probably shouldn't do this. Except for maybe asserting email uniqueness.